src/hashset.c

```diff
@@ -438,48 +438,61 @@ static inline int cursorIsLessThan(size_t a, size_t b) {
     return rev(a) < rev(b);
 }
 
-/* Rehashes one bucket. */
+/* Rehashes buckets, following potential probe chains. This means that between
+ * rehash steps, any element that hashes to an already-rehashed bucket has been
+ * rehashed, even if it was moved to a different bucket due to open addressing.
+ * Knowing this allows scan to skip already-rehashed buckets. */
 static void rehashStep(hashset *s) {
     assert(hashsetIsRehashing(s));
     size_t idx = s->rehash_idx;
-    bucket *b = &s->tables[0][idx];
-    int pos;
-    for (pos = 0; pos < ELEMENTS_PER_BUCKET; pos++) {
-        if (!isPositionFilled(b, pos)) continue; /* empty */
-        void *element = b->elements[pos];
-        uint8_t h2 = b->hashes[pos];
-        /* Insert into table 1. */
-        uint64_t hash;
-        /* When shrinking, it's possible to avoid computing the hash. We can
-         * just use idx has the hash, but only if we know that probing didn't
-         * push this element away from its primary bucket, so only if the
-         * bucket before the current one hasn't ever been full. */
-        if (s->bucket_exp[1] <= s->bucket_exp[0] &&
-            !s->tables[0][prevCursor(idx, expToMask(s->bucket_exp[0]))].everfull) {
-            hash = idx;
-        } else {
-            hash = hashElement(s, element);
+    uint8_t prev_bucket_everfull = s->tables[0][prevCursor(idx, expToMask(s->bucket_exp[0]))].everfull;
+    while (1) {
+        bucket *b = &s->tables[0][idx];
+        if (b->presence) {
```

Comment:
To avoid adding a big indented block, we could use goto.

```suggestion
        if (!b->presence) goto bucket_done;
```

Reply:
With your helper function idea I was able to make this `if (!b->presence) return;`